vsphere/virtual_machine_config_structure.go

```diff
@@ -508,6 +509,117 @@ func flattenExtraConfig(d *schema.ResourceData, opts []types.BaseOptionValue) er
 	return d.Set("extra_config", ec)
 }
 
+// expandVAppConfig reads in all the vapp key/value pairs and returns
+// the appropriate VmConfigSpec.
+//
+// We track changes to keys to determine if any have been removed from
+// configuration - if they have, we add them with an empty value to ensure
+// they are removed from vAppConfig on the update.
+func expandVAppConfig(d *schema.ResourceData) types.BaseVmConfigSpec {
+	if !d.HasChange("vapp") {
+		return nil
+	}
+
+	var (
+		props  []types.VAppPropertySpec
+		allIds sort.StringSlice
+	)
+
+	old, new := d.GetChange("vapp")
+
+	getVAppProps := func(val interface{}) map[string]interface{} {
+		vApps := val.([]interface{})
+		if vApps != nil && len(vApps) > 0 && vApps[0] != nil {
+			vApp := vApps[0].(map[string]interface{})
+			if props, ok := vApp["properties"].(map[string]interface{}); ok {
+				return props
+			}
+		}
+		return make(map[string]interface{})
+	}
+
+	oldMap := getVAppProps(old)
+	newMap := getVAppProps(new)
+
+	for k := range oldMap {
+		allIds = append(allIds, k)
+	}
+
+	for k := range newMap {
+		if _, ok := oldMap[k]; !ok {
+			allIds = append(allIds, k)
+		}
+	}
+	sort.Sort(allIds)
+
+	for i, k := range allIds {
+		_, isOld := oldMap[k]
+		_, isNew := newMap[k]
+
+		if isNew {
+			prop := types.VAppPropertySpec{
+				ArrayUpdateSpec: types.ArrayUpdateSpec{
+					Operation: types.ArrayUpdateOperationEdit,
+				},
+				Info: &types.VAppPropertyInfo{
+					Key:   int32(i) + 1,
+					Id:    k,
+					Value: newMap[k].(string),
+				},
+			}
+			props = append(props, prop)
+		} else if isOld {
+			prop := types.VAppPropertySpec{
+				ArrayUpdateSpec: types.ArrayUpdateSpec{
+					Operation: types.ArrayUpdateOperationEdit,
+				},
+				Info: &types.VAppPropertyInfo{
+					Key:   int32(i) + 1,
+					Id:    k,
+					Value: "",
+				},
+			}
+			props = append(props, prop)
+		}
+	}
+
+	// Done!
+	return &types.VmConfigSpec{
+		Property: props,
+	}
+}
+
+// flattenVAppConfig reads in the vAppConfig from a running virtual machine
+// and *only* sets the keys in vapp that we know about.
+func flattenVAppConfig(d *schema.ResourceData, config types.BaseVmConfigInfo) error {
+	if config == nil {
+		return nil
+	}
+	props := config.GetVmConfigInfo().Property
+	if len(props) < 1 {
+		// No props to read is a no-op
+		return nil
+	}
+	vac := make(map[string]interface{})
+	vApp := d.Get("vapp").([]interface{})
+	if len(vApp) > 0 && vApp[0] != nil {
```

Comment:
As mentioned in the comments before the change request, I think we need to add config values that have non-zero values set on them as well, that way, on the next apply, they get zeroed out and any defaults apply. That way, removing a key always removes the corresponding value from config, even if someone adds OOB changes.